docs/prover/approx/hashing.md

```diff
@@ -0,0 +1,182 @@
+The entries of mapping `m` are spread out over storage; their locations are computed as the hash of the mapping's storage slot and the key at which the mapping is being accessed; thus the storage slot used for the entry of `m` under key `k` is computed as `hash(2, k)`.
+
+We can see that non-collision of hashes is essential for storage integrity. E.g., if `hash(1) + j` was equal to `hash(2, k)` then the operations on `a` an `m` would interfere with each other.
+
+Also note that the initial storage slots are reserved, i.e., we make sure that no hash value ends up colliding with slots 0 to 10000.
+
+
+## Modeling the Keccak Function (bounded case)
+
+The Certora Prover models the Keccak hash function as an arbitrary function that is _injective with large gaps_.
+That means that if `x != y` then `hash(x) != hash(y)`, but also that for all additive offsets `o` that actually occur in the program `hash(x) + o != hash(y)`.
+
+These constraints are enough for the solidity storage model to work as expected. However when hashes are compared, they might show different behavior from the actual Keccak function (e.g. `hash(x) > hash(y)` is always true or false for given `x` and `y`, but in our formulas, the SMT solver is free to choose a function for `hash` that makes the formula satisfiable.). 
+We have not observed a practical use case yet where the numeric values of the hash function play a role, thus we chose this modeling for tractability reasons.
+
+
+### Example (Imprecision of Modeling)
+
+Whichever distinct values we chose for `x` and `y` in the example below, on the real Keccak function one rule would be violated and one rule would not. In the modeling of the Certora Prover, both rules are violated, since the prover is allowed to "invent" a hash function for each rule and will choose one that violates the property.
+
+```solidity
+// CVL:
+methods { hash(uint) returns (uint) envfree; }
+
+definition x() : uint = 12345678
+definition y() : uint = 87654321
+
+rule hashXLowerOrEqualToHashY {
+	assert hash(x()) <= hash(y());
+}
+
+rule hashXLargerThanHashY {
+	assert hash(x()) > hash(y());
+}
+
+// solidity:
+contract C {
+	function hash(uint x) public returns (bytes32) {
+		return keccak256(x);
+	}
+}
+
+```
+
+
+<!---
+[comment]: # We should also clearly explain the surprising thing the user might see. A clear example of a rule that should pass and doesn't, or a rule that shouldn't pass and does, and an explanation of why.
+--->
+
+## Hashing of Unbounded Data
+
+In the discussion so far we only considered hashes of data whose length is already known before program execution (e.g. a `uint` variable always has 256 bit). Hashing of unbounded data (typically unbounded arrays, like `bytes`, `uint[]`, etc.) requires some extra measures, since their implementation requires loops and the the Certora Prover internally eliminates all loops to other constructs in order to achieve better tractability.
```

Comment:
256 bit**s**

Reply:
strike "to other constructs"

Reply:
possibly add a link to the page on loops
```
{doc}`internally eliminates loops <loops>`
```

Reply:
done